### PR TITLE
Make ShaderLanguage's parser recognize sampler passed in from array

### LINKED ISSUE #94447: Passing sampler from array as parameter to non-built-in function throws error.
### Tested versions

Reproducible in 4.3 beta3 and 4.2.1 stable.

### System information

Godot v4.3.beta3.mono - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 2060 (NVIDIA; 32.0.15.5612) - Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz (16 Threads)

### Issue description

When writing a shader function that takes a sampler as an argument, calling that function while indexing into an array of samplers results in an error. The shader seems to still work fine, as the error only results in continuing the current loop iteration of the shader parsing. The error in question is triggered at line 5550 of servers/rendering/shader_language.cpp.

It seems that the shader parser expects sampler function arguments to only ever be referred to by variable identifiers representing a single sampler and not elements of an array of samplers. In my experience this error only ever triggers for user-defined functions, not built-ins.

### Steps to reproduce

```glsl
shader_type canvas_item;

uniform sampler2D textures[1];

vec4 sampleTex(sampler2D tex, vec2 uv) {
    return texture(tex, uv);
}

void fragment() {
    COLOR = sampleTex(textures[0], UV); 
}
```

### Minimal reproduction project (MRP)

https://github.com/operation404/ShaderSamplerArrayBugMRP

### PR DESCRIPTION
**Pull request by jsjtxietian**
Fixes https://github.com/godotengine/godot/issues/94447

### CODE REVIEW

File: servers/rendering/shader_language.cpp [hunk 1]
```diff
@@ -5549,9 +5549,22 @@ ShaderLanguage::Node *ShaderLanguage::_parse_expression(BlockNode *p_block, cons
...
+											varname = vn->name;
+										} else if (n->type == Node::NODE_TYPE_ARRAY) {
+											ArrayNode *an = static_cast<ArrayNode *>(n);
+											if (is_sampler_type(an->datatype_cache)) {
+												varname = an->name;
+											} else {
+												ERR_CONTINUE("The argument passed in is not a sampler.");
+											}
+										} else {
+											ERR_CONTINUE("The argument passed in is not a sampler.");
```
Comment (Chaosus) [inline servers/rendering/shader_language.cpp:5565]: `ERR_CONTINUE` is not intended for such usage, there is not a single occurrence of it with an error string as an argument across the engine. You should use `ERR_CONTINUE_MSG` if you really need it.

Comment (jsjtxietian) [reply servers/rendering/shader_language.cpp:5565]: Sorry I misunderstood it.

File: servers/rendering/shader_language.cpp [hunk 2]
```diff
@@ -5549,9 +5549,22 @@ ShaderLanguage::Node *ShaderLanguage::_parse_expression(BlockNode *p_block, cons
...
+											ArrayNode *an = static_cast<ArrayNode *>(n);
+											if (is_sampler_type(an->datatype_cache)) {
+												varname = an->name;
+											} else {
+												ERR_CONTINUE_MSG(true, "The argument passed in is not a sampler.");
+											}
+										} else {
+											ERR_CONTINUE_MSG(true, "The argument passed in is not a sampler.");
+										}
+
```
Comment (Chaosus) [inline servers/rendering/shader_language.cpp:5567]:
```suggestion
										StringName varname;
										if (n->type == Node::NODE_TYPE_VARIABLE) {
											VariableNode *vn = static_cast<VariableNode *>(n);
											varname = vn->name;
										} else if (n->type == Node::NODE_TYPE_ARRAY) {
											ArrayNode *an = static_cast<ArrayNode *>(n);
											varname = an->name;
										}
```

Type checking occurs before that code, so there is no reason to check it detailed and emit errors here.

**Review by Chaosus — CHANGES_REQUESTED** [review]

File: servers/rendering/shader_language.cpp (hunk 2 above)
Comment (jsjtxietian) [reply servers/rendering/shader_language.cpp:5567]: Thanks!

File: servers/rendering/shader_language.cpp [hunk 3]
```diff
@@ -5549,9 +5549,16 @@ ShaderLanguage::Node *ShaderLanguage::_parse_expression(BlockNode *p_block, cons
 									}
 									if (is_sampler_type(call_function->arguments[i].type)) {
 										//let's see where our argument comes from
```
Comment (Chaosus) [inline servers/rendering/shader_language.cpp:5551]:
```suggestion
										// Let's see where our argument comes from.
```

File: servers/rendering/shader_language.cpp [hunk 4]
```diff
@@ -5549,9 +5549,16 @@ ShaderLanguage::Node *ShaderLanguage::_parse_expression(BlockNode *p_block, cons
...
-										ERR_CONTINUE(n->type != Node::NODE_TYPE_VARIABLE); //bug? this should always be a variable
-										VariableNode *vn = static_cast<VariableNode *>(n);
-										StringName varname = vn->name;
+
```
Comment (Chaosus) [inline servers/rendering/shader_language.cpp:5552]:
```suggestion
```

**Review by Chaosus — CHANGES_REQUESTED** [review]

**Review by Chaosus — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**